internal/v1/handler.go

```diff
@@ -711,3 +711,195 @@ func buildCustomizations(cust *Customizations) *composer.Customizations {
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		return echo.NewHTTPError(http.StatusBadRequest, "Cloning a compose is only available for aws composes")
+	}
+
+	if resp == nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, "Something went wrong creating the clone")
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusCreated {
+		var cError composer.Error
+		err = json.NewDecoder(resp.Body).Decode(&cError)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Unable to parse error returned by image-builder-composer service")
+		}
+		return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("image-builder-composer service returned an error: %s", cError.Reason))
+	}
+
+	var cloneResponse composer.CloneComposeResponse
+	err = json.NewDecoder(resp.Body).Decode(&cloneResponse)
+	if err != nil {
+		logrus.Errorf("Unable to decode CloneComposeResponse: %v", err)
+		return err
+	}
+
+	err = h.server.db.InsertClone(composeId, cloneResponse.Id.String(), rawCR)
+	if err != nil {
+		logrus.Errorf("Error inserting clone into db: %v", err)
+		return err
+	}
+
+	return ctx.JSON(http.StatusCreated, CloneResponse{
+		Id: cloneResponse.Id.String(),
+	})
+}
+
+func (h *Handlers) GetCloneStatus(ctx echo.Context, id string) error {
+	idHeader, err := getIdentityHeader(ctx)
+	if err != nil {
+		return err
+	}
+
+	cloneEntry, err := h.server.db.GetClone(id, idHeader.Identity.OrgID)
+	if err != nil {
+		if errors.As(err, &db.CloneNotFoundError) {
+			return echo.NewHTTPError(http.StatusNotFound, err)
+		}
+		return err
+	}
+	if cloneEntry == nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "Requested clone cannot be found")
+	}
+
+	resp, err := h.server.client.CloneStatus(id)
+	if err != nil {
+		logrus.Errorf("Error requesting clone status: %v", err)
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		var cErr composer.Error
+		err = json.NewDecoder(resp.Body).Decode(&cErr)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Unable to parse composer error")
+		}
+		return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Unable to create clone job: %v", cErr.Reason))
+	}
+
+	var cloudStat composer.CloneStatus
+	err = json.NewDecoder(resp.Body).Decode(&cloudStat)
+	if err != nil {
+		logrus.Errorf("Unable to decode clone status: %v", err)
+		return err
+	}
+
+	return ctx.JSON(http.StatusOK, UploadStatus{
+		Status:  UploadStatusStatus(cloudStat.Status),
+		Type:    UploadTypes(cloudStat.Type),
+		Options: cloudStat.Options,
+	})
+}
+
+func (h *Handlers) GetComposeClones(ctx echo.Context, composeId string, params GetComposeClonesParams) error {
+	err := h.canUserAccessComposeId(ctx, composeId)
+	if err != nil {
+		return err
+	}
+
+	idHeader, err := getIdentityHeader(ctx)
+	if err != nil {
+		return err
+	}
+
+	imageType, err := h.server.db.GetComposeImageType(composeId, idHeader.Identity.OrgID)
```

Comment:
I wonder if this check is even needed. Querying a compose other than an aws one does not harm. :shrug:

Reply:
true